### PR TITLE
use the Resource Manager

### PR DESCRIPTION
**Pull request by marten-seemann**
Depends on https://github.com/libp2p/go-libp2p-core/pull/229.

TODO: write tests.

### CODE REVIEW

**Review by vyzo — COMMENTED** [review]
this is looking quite ergonomic and nice!

File: upgrader.go [hunk 1]
```diff
@@ -106,12 +107,20 @@ func (u *Upgrader) Upgrade(ctx context.Context, t transport.Transport, maconn ma
...
 		if err := maconn.Close(); err != nil {
-			log.Errorf("failed to close connection with peer %s and addr %s; err: %s",
-				p.Pretty(), maconn.RemoteMultiaddr(), err)
+			log.Errorw("failed to close connection", "peer", p, "addr", maconn.RemoteMultiaddr(), "error", err)
 		}
 		return nil, fmt.Errorf("gater rejected connection with peer %s and addr %s with direction %d",
 			sconn.RemotePeer().Pretty(), maconn.RemoteMultiaddr(), dir)
 	}
+	if p == "" && connScope != nil {
+		if err := connScope.SetPeer(sconn.RemotePeer()); err != nil {
```
Comment (vyzo) [inline upgrader.go:116]: we need a Done in the error path, as we own it here.

File: listener.go [hunk 2]
```diff
@@ -86,13 +87,25 @@ func (l *listener) handleIncoming() {
...
-
 			if err := maconn.Close(); err != nil {
-				log.Warnf("failed to incoming connection rejected by gater; err: %s", err)
+				log.Warnf("failed to close incoming connection rejected by gater: %s", err)
 			}
 			continue
 		}
 
+		var connScope network.ConnectionManagementScope
+		if l.rcmgr != nil {
```
Comment (vyzo) [inline listener.go:97]: nil check unnecessary?

File: upgrader.go [hunk 3]
```diff
@@ -45,12 +45,13 @@ type Upgrader struct {
 }
 
 // UpgradeListener upgrades the passed multiaddr-net listener into a full libp2p-transport listener.
-func (u *Upgrader) UpgradeListener(t transport.Transport, list manet.Listener) transport.Listener {
+func (u *Upgrader) UpgradeListener(t transport.Transport, list manet.Listener, rcmgr network.ResourceManager) transport.Listener {
```
Comment (vyzo) [inline upgrader.go:48]: shouldnt it have its own rcmgr?

Comment (marten-seemann) [reply upgrader.go:48]: Not sure. Would that be the better design?

File: listener.go (hunk 2 above)
Comment (marten-seemann) [reply listener.go]: Fixed.

File: upgrader.go (hunk 3 above)
Comment (vyzo) [reply upgrader.go:48]: hrm, maybe. Shouldnt everything have the rcmgr?

Comment (marten-seemann) [reply upgrader.go]: Done.

File: upgrader.go (hunk 1 above)
Comment (vyzo) [reply upgrader.go:163]: so where is this getting Done? The caller is responsible?

**Review by vyzo — APPROVED** [review]
LGTM modulo the comment.

File: upgrader.go (hunk 1 above)
Comment (marten-seemann) [reply upgrader.go:163]: In the swarm: https://github.com/libp2p/go-libp2p-swarm/pull/308/files#diff-decb5d093ca975d319efcad30e802c35efa8f38ede82777bbd206239ca01b9bdR97

**Pull request merged**